In [14]:
import itertools
import pandas as pd
import numpy as np

Your task is to develop two functions, one for each scenario, where users can input values for m, n, and x to compute the required probability.

# Scenario 1

Calculating the probability of rolling m dice, each with n sides, and having the sum of the dice be greater than x. This calculation should only be implemented for m = 1 through 4 and n is at most 20.

In [28]:
def probability_sum_greater_than_x(m, n, x):
    if(x > 80):
        # The maximum sum of 4d20 is 80
        print(f"Invalid input. x must be at most 80.")
        return 000
    
    elif((1 <= m <= 4) and (n <= 20)):
        # List out all possible rolls in range 1:n+1 (since range is exclusive) for m dice
        rolls = list(itertools.product(range(1, (n + 1)), repeat=m))
        # Find the sum of all possible rolls
        sums = np.array([sum(roll) for roll in rolls])
        # Find the proportion of those sums that are greater than the target x
        return np.mean(sums > x) 
    
    else:
        # If the input is invalid (m < 1, m > 4, n > 20), return 000
        print(f"Invalid input. m must be between 1 and 4, and n must be at most 20.")
        return 000

# FOR TESTING PURPOSES
# prob_3d4 = probability_sum_greater_than_x(4, 15, 13)
# Should output 0.9859

# print(f"The probability of rolling 3d4 sum > 9: {prob_3d4:.4f}")


# Scenario 2

Determining the probability of rolling m dice, each with n sides, and having at least one die show a value greater than or equal to x.

In [16]:
def probability_of_at_least_one_x_or_greater(m, n, x):
    # List out all possible rolls
    rolls = list(itertools.product(range(1, n+1), repeat=m))
    # Find the number of rolls with at least one dice >= x 
    num_above_x = sum(1 for roll in rolls if any(dice >=x for dice in roll))
    # Find the overall probability of a num_above_x
    probability = num_above_x/len(rolls)
    return probability

# Walkthrough Example

Welcome Red Dragon Inn. We are excited to announce our new tools to help you become a master in D&D.

We are introducing two new challenges: _Master of Dice_ and _The Greater_ with these new challenges you will be able to master new abilities, increase your confidence and dominate youre D&D endeavors! Let's see how it works.

**Challenge:** Let's do a 1D20 to get 17 or more!

In [29]:
prob_1d20 = probability_of_at_least_one_x_or_greater(1,20,17)
print(f"Probability of rolling 1D20 and getting 17 or more {prob_1d20:.2f}")

Probability of rolling 1D20 and getting 17 or more 0.20


1. In this example, we want to calculate the probability of rolling 1 d20 and the value being greater or equal to 17. There are 4 options to roll successfully (17, 18, 19, 20). 4/20 gives us a 20% probability.\
\
**You get a** 1 in 5 chance to get it!



**Challenge:** Let's do 3D4 and get a result of 10 or more!

In [30]:
prob3d4 = probability_sum_greater_than_x(3,4,9)
print(f"Probability of rolling 3D4 and getting a sum of 10 or more {prob3d4:.4f}")

Probability of rolling 3D4 and getting a sum of 10 or more 0.1562


2. Here we want to roll 3 d4 and get a sum greater than or equal to 10. There are 64 total possibilities 
(4 sides)^3 rolls and we will only succeed if we get a 10, 11 or 12. There are 6 ways we can roll a 10:
(2, 4, 4)
(4, 2, 4)
(4, 4, 2)
(3, 3, 4)
(3, 4, 3)
(4, 3, 3)
There are only 3 ways of rolling 11:
(4, 4, 3)
(4, 3, 4)
(3, 4, 4)
And only 1 way of rolling a 12:
(4, 4, 4)
This means that we succeed 10 times out of 64 total possible outcomes. P(3d4 >= 10) = 10/64 = 15.625%\
\

**You get** a 15.625% chance, now you can decide better!

**Challenge:** Let's see a 4D6 and have at least one dice show a 6!

In [19]:
prob4d6 = probability_of_at_least_one_x_or_greater(4,6,6)
print(f"{prob4d6:.4f}")

0.5177


3. Rolling 4 d6 and having at least one dice showing 6: In this scenario, each dice has a 1 - (5/6)^4 chance to land on 6.

**You get** a 0.5177 (51.77%) chance to get one of them to land on  6!